File: cores/typescript/src/mappers/axrom.ts
// Mapper 7 (AxROM). Port of cores/c/src/mappers/axrom.c.

import { Mapper, Mirroring } from '../mapper';

export class Axrom extends Mapper {
    private static readonly PrgBankSize = 32768;
    private static readonly ChrSize = 8192;

    private _prgRom: Uint8Array;
    private _prgSize: number;
    private _chr: Uint8Array;
    private _chrSize: number;
    private _chrIsRam: boolean;
    private _hasBattery: boolean;
    private _prgBank: number;
    private _mirrorNt: number;

    constructor(prg: Uint8Array, prgSize: number, chr: Uint8Array | null, chrSize: number, mirroring: number, hasBattery: boolean) {
        super();
        this.mapperNum = 7;
        this._prgSize = prgSize;
        this._prgRom = new Uint8Array(prgSize);
        if (prgSize > 0) this._prgRom.set(prg.subarray(0, prgSize));
        if (chrSize > 0) {
            this._chrSize = chrSize;
            this._chr = new Uint8Array(chrSize);
            this._chr.set(chr!.subarray(0, chrSize));
            this._chrIsRam = false;
        } else {
            this._chrSize = Axrom.ChrSize;
            this._chr = new Uint8Array(Axrom.ChrSize);
            this._chrIsRam = true;
        }
        this._hasBattery = hasBattery;
        this._prgBank = 0;
        this._mirrorNt = 0;
    }

    private prgBankCount(): number {
        const c = Math.floor(this._prgSize / Axrom.PrgBankSize);
        return c > 0 ? c : 1;
    }

    readPrg(addr: number): number {
        if (addr < 0x8000) return 0x00;
        const local = addr - 0x8000;
        const count = this.prgBankCount();
        const bank = this._prgBank % count;
        const idx = bank * Axrom.PrgBankSize + local;
        if (idx >= this._prgSize) return 0x00;
        return this._prgRom[idx];
    }

    writePrg(addr: number, value: number): void {
        if (addr < 0x8000) return;
        this._mirrorNt = (value >> 4) & 1;
        this._prgBank = value & 0x07;
    }

    readChr(addr: number): number {
        const sz = this._chrSize;
        if (sz === 0) return 0x00;
        return this._chr[addr % sz];
    }

    writeChr(addr: number, value: number): void {
        if (!this._chrIsRam) return;
        const sz = this._chrSize;
        if (sz === 0) return;
        this._chr[addr % sz] = value;
    }

    mirrorMode(): number { return this._mirrorNt === 0 ? Mirroring.SingleScreen0 : Mirroring.SingleScreen1; }
    chrIsRam(): boolean { return this._chrIsRam; }
    hasBattery(): boolean { return this._hasBattery; }

    saveState(buf: Uint8Array | null): number {
        const total = 2 + (this._chrIsRam ? this._chrSize : 0);
        if (buf === null) return total;
        let p = 0;
        buf[p++] = this._prgBank;
        buf[p++] = this._mirrorNt;
        if (this._chrIsRam && this._chrSize > 0) {
            buf.set(this._chr.subarray(0, this._chrSize), p);
            p += this._chrSize;
        }
        return p;
    }

    loadState(buf: Uint8Array, len: number): boolean {
        const need = 2 + (this._chrIsRam ? this._chrSize : 0);
        if (len < need) return false;
        let p = 0;
        this._prgBank = buf[p++];
        this._mirrorNt = buf[p++];
        if (this._chrIsRam && this._chrSize > 0) {
            this._chr.set(buf.subarray(p, p + this._chrSize));
            p += this._chrSize;
        }
        return true;
    }
}

File: cores/csharp/src/Mappers/Axrom.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// Mapper 7 (AxROM). Port of src/mappers/axrom.rs.
public sealed class Axrom : Mapper
{
    private const uint PrgBankSize = 32768;
    private const uint ChrSize = 8192;

    private byte[] _prgRom;
    private uint _prgSize;
    private byte[] _chr;
    private uint _chrSize;
    private bool _chrIsRam;
    private bool _hasBattery;
    private byte _prgBank;
    private byte _mirrorNt;

    public Axrom(byte[] prg, uint prgSize, byte[] chr, uint chrSize, Mirroring mirroring, bool hasBattery)
    {
        MapperNum = 7;
        _prgSize = prgSize;
        _prgRom = new byte[prgSize];
        if (prgSize > 0) System.Array.Copy(prg, _prgRom, (int)prgSize);
        if (chrSize > 0)
        {
            _chrSize = chrSize;
            _chr = new byte[chrSize];
            System.Array.Copy(chr, _chr, (int)chrSize);
            _chrIsRam = false;
        }
        else
        {
            _chrSize = ChrSize;
            _chr = new byte[ChrSize];
            _chrIsRam = true;
        }
        _hasBattery = hasBattery;
        _prgBank = 0;
        _mirrorNt = 0;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private uint PrgBankCount() => _prgSize / PrgBankSize > 0 ? _prgSize / PrgBankSize : 1;

    public override byte ReadPrg(ushort addr)
    {
        if (addr < 0x8000) return 0x00;
        uint local = (uint)(addr - 0x8000);
        uint count = PrgBankCount();
        uint bank = _prgBank % count;
        uint idx = bank * PrgBankSize + local;
        if (idx >= _prgSize) return 0x00;
        return _prgRom[idx];
    }

    public override void WritePrg(ushort addr, byte value)
    {
        if (addr < 0x8000) return;
        _mirrorNt = (byte)((value >> 4) & 1);
        _prgBank = (byte)(value & 0x07);
    }

    public override byte ReadChr(ushort addr)
    {
        uint sz = _chrSize;
        if (sz == 0) return 0x00;
        return _chr[addr % sz];
    }

    public override void WriteChr(ushort addr, byte value)
    {
        if (!_chrIsRam) return;
        uint sz = _chrSize;
        if (sz == 0) return;
        _chr[addr % sz] = value;
    }

    public override Mirroring MirrorMode() => _mirrorNt == 0 ? Mirroring.SingleScreen0 : Mirroring.SingleScreen1;
    public override bool ChrIsRam() => _chrIsRam;
    public override bool HasBattery() => _hasBattery;

    public override uint SaveState(byte[] buf)
    {
        uint total = 2 + (_chrIsRam ? _chrSize : 0);
        if (buf == null) return total;
        int p = 0;
        buf[p++] = _prgBank;
        buf[p++] = _mirrorNt;
        if (_chrIsRam && _chrSize > 0)
        {
            System.Array.Copy(_chr, 0, buf, p, (int)_chrSize);
            p += (int)_chrSize;
        }
        return (uint)p;
    }

    public override bool LoadState(byte[] buf, uint len)
    {
        uint need = 2 + (_chrIsRam ? _chrSize : 0);
        if (len < need) return false;
        int p = 0;
        _prgBank = buf[p++];
        _mirrorNt = buf[p++];
        if (_chrIsRam && _chrSize > 0)
        {
            System.Array.Copy(buf, p, _chr, 0, (int)_chrSize);
            p += (int)_chrSize;
        }
        return true;
    }
}
